keps/sig-storage/1790-recover-resize-failure/README.md

```diff
@@ -102,64 +103,68 @@ reduces the PVC request size, for both CSI and in-tree plugins they are designed
 
-AllocatedResources is not volume size but more like whatever user has requested and towards which resize-controller was working to reconcile. It is possible that user has requested smaller size since then but that does not changes the fact that resize-controller has already tried to expand to AllocatedResources and might have partially succeeded. So AllocatedResources is maximum user requested size for this volume and does not reflect actual volume size of the PV and hence is not recorded in `pvc.Status.Capacity`.
-
-This also falls inline with how in-place update of Pod is handled(https://github.com/kubernetes/enhancements/pull/1342/files) and allocated resources is recorded in `pod.Spec.Containers[i].ResourcesAllocated`.
+*Note:* This proposal expects that users can not modify `pvc.Status` directly and cheat quota system. In general this should be fine because users should not have access to edit status of almost anything. Link to discussion on slack - https://kubernetes.slack.com/archives/CJUQN3E4T/p1620059624022100
 
 #### User flow stories
 
-###### Case 1 (controller+node expandable):
+##### Case 0 (default PVC creation):
+- User creates a 10Gi PVC by setting - `pvc.spec.resources.requests["storage"] = "10Gi"`.
+- API server sets `pvc.Status.AllocatedResources` to "10Gi" on creation.
 
+##### Case 1 (controller+node expandable):
 - User increases 10Gi PVC to 100Gi by changing - `pvc.spec.resources.requests["storage"] = "100Gi"`.
-- API Server sets `pvc.spec.allocatedResources` to 100Gi via pvc storage strategy.
-- Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `90Gi` to used quota.
+- `pvc.Status.AllocatedResources` still reports `10Gi`.
+- Quota controller uses `max(pvc.Status.AllocatedResources, pvc.Spec.Resources)` and adds `90Gi` to used quota.
+- Expansion controller starts expanding the volume and sets `pvc.Status.AllocatedResources` to `100Gi`.
 - Expansion to 100Gi fails and hence `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 10Gi.
 - User requests size to 20Gi.
-- Quota controller sees no change in storage usage by the PVC because `pvc.Spec.AllocatedResources` is `100Gi`.
+- Expansion controller notices that `pvc.Spec.Resources` < `pvc.Status.AllocatedResources` (meaning user tried to reduce size).
+- Expansion controller fetches actual size of the volume using `ControllerGetVolume` CSI RPC call.
```

Comment:
How do we ensure that there isn't an expansion operation still ongoing in the system?

Reply:
I think it is tricky to accurately determine if expansion is ongoing in the system. One tricky case that comes to my mind is - what if expansion controller starts expanding but before it can update success/failure in PVC, it crashed. In that case - when user reduces the size, code will assume no expansion is in progress and hence no quota will be returned back to the user (because I presume - if we are checking for resizing in-progress, then we will want to reduce `pvc.Status.AllocatedResources` only if no expansion is progress in the system).

Alternatively - we could adjust `pvc.Status.AllocatedResources` *after* the control-plane expansion. One limitation with that is - then for node only expandable volumes, we won't be able to accurately adjust `pvc.Status.AllocatedResources` because an expansion on the node might still be in progress.

Reply:
The design was updated to adjust `pvc.Status.AllocatedResources` *after* both control and node side expansion finishes. I think this should take care of in-progress expansion operations.